### ISSUE TITLE
Can't run hello-world

### ISSUE THREAD

**Issue by tremblerz** [QISKit/qiskit-sdk-py#309]
Experiencing error with Quantum emoticon program
## Expected Behavior
I am not really sure about the expected behavior. Given that it is a one of the trivial program, it should not be a problem for the developers.

## Current Behavior
In this [file](https://github.com/QISKit/qiskit-tutorial/blob/master/0_hello_world/quantum_emoticon.ipynb) 3rd block is throwing the following error
```

---------------------------------------------------------------------------
ResultError                               Traceback (most recent call last)
<ipython-input-1-72b078066956> in <module>()
     31 # run and get results
     32 results = qp.execute(["smiley_writer"], backend='ibmqx5', shots=1024)
---> 33 stats = results.get_counts("smiley_writer")

/anaconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit-0.5.0-py3.6-macosx-10.7-x86_64.egg/qiskit/_result.py in get_counts(self, name)
    231         """
    232         try:
--> 233             return self.get_data(name)['counts']
    234         except KeyError:
    235             raise QISKitError('No counts for circuit "{0}"'.format(name))

/anaconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit-0.5.0-py3.6-macosx-10.7-x86_64.egg/qiskit/_result.py in get_data(self, name)
    202             exception = self._result['result']
    203             if isinstance(exception, BaseException):
--> 204                 raise exception
    205             else:
    206                 raise QISKitError(str(exception))

/anaconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit-0.5.0-py3.6-macosx-10.7-x86_64.egg/qiskit/_jobprocessor.py in _job_done_callback(self, future)
     97     def _job_done_callback(self, future):
     98         try:
---> 99             result = future.result()
    100         except Exception as ex:  # pylint: disable=broad-except
    101             result = Result({'job_id': '0', 'status': 'ERROR',

/anaconda3/envs/QISKitenv/lib/python3.6/concurrent/futures/_base.py in result(self, timeout)
    423                 raise CancelledError()
    424             elif self._state == FINISHED:
--> 425                 return self.__get_result()
    426 
    427             self._condition.wait(timeout)

/anaconda3/envs/QISKitenv/lib/python3.6/concurrent/futures/_base.py in __get_result(self)
    382     def __get_result(self):
    383         if self._exception:
--> 384             raise self._exception
    385         else:
    386             return self._result

/anaconda3/envs/QISKitenv/lib/python3.6/concurrent/futures/thread.py in run(self)
     54 
     55         try:
---> 56             result = self.fn(*self.args, **self.kwargs)
     57         except BaseException as exc:
     58             self.future.set_exception(exc)

/anaconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit-0.5.0-py3.6-macosx-10.7-x86_64.egg/qiskit/_jobprocessor.py in run_backend(q_job)
     52                 circuit['compiled_circuit'] = compiled_circuit
     53     backend = qiskit.backends.get_backend_instance(backend_name)
---> 54     return backend.run(q_job)
     55 
     56 

/anaconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit-0.5.0-py3.6-macosx-10.7-x86_64.egg/qiskit/backends/_qeremote.py in run(self, q_job)
     98                                    hpc=hpc)
     99         if 'error' in output:
--> 100             raise ResultError(output['error'])
    101 
    102         logger.debug('Running on remote backend %s with job id: %s',

ResultError: QASM_NOT_VALID: Error parsing QASM. Error parsing qasm number 0. Gates after a measure are blocked

```

## Possible Solution
Not sure about it.

## Steps to Reproduce (for bugs)
Just try to execute `hello-world` example

## Context
I was trying to getting started with some development around it.

## Your Environment
* Version used: qiskit-0.5.0
* Environment name and version: conda-4.4.10, Python 3.6.1
* MacOS Sierra

**Comment by 1ucian0** [comment]
thanks @tremblerz to let us know about the issue. Here is a possible action plan:

1. Did Quantum emoticon program worked in 4.9.0? Is this a regression?
2. If 1. yes. Is the mapper doing what it suppose to do?
3. if 2. yes. The compiler process needs to be smarter and adds a barrier before measure to avoid further operation on the measured qbits.

**Comment by indisoluble** [comment]
Hi @tremblerz,

Few days ago I pointed this same problem out [here](https://github.com/QISKit/qiskit-tutorial/issues/88) and @rraymondhp provided the following answer:

> Hi,
> Thank you for pointing this. I also got the same error when running the notebook. It is due to a bug in compiler that swap measurement to the middle of the computation but the devices do not support such operation yet. I confirmed that I can avoid the error by inserting barrier before measurement, like:
> 
> //put barrier before measurement
> qc.barrier(qr)
> //measure
> for j in range(16):
>     qc.measure(qr[j], cr[j])
> 
> //run and get results
> results = qp.execute(["smiley_writer"], backend='ibmqx5', shots=1024, timeout=240)
> stats = results.get_counts("smiley_writer")
> 
> I cannot run the real device now because it is under maintenance though.

**Comment by ajavadia** [comment]
Fixed via #347
